hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ReplicationManager.java

```diff
@@ -0,0 +1,448 @@
+
+  /**
+   * Constructs ReplicationManager instance with the given configuration.
+   *
+   * @param conf OzoneConfiguration
+   * @param containerManager ContainerManager
+   * @param containerPlacement PlacementPolicy
+   * @param eventPublisher EventPublisher
+   */
+  @SuppressWarnings("parameternumber")
+  public ReplicationManager(final ConfigurationSource conf,
+             final ContainerManager containerManager,
+             final PlacementPolicy containerPlacement,
+             final EventPublisher eventPublisher,
+             final SCMContext scmContext,
+             final SCMServiceManager serviceManager,
+             final NodeManager nodeManager,
+             final Clock clock,
+             final SCMHAManager scmhaManager,
+             final Table<ContainerID, MoveDataNodePair> moveTable)
+             throws IOException {
+    this.containerManager = containerManager;
+    this.scmContext = scmContext;
+    this.rmConf = conf.getObject(ReplicationManagerConfiguration.class);
+    this.running = false;
+    this.clock = clock;
+    this.containerReport = new ReplicationManagerReport();
+    this.metrics = null;
+    this.waitTimeInMillis = conf.getTimeDuration(
+        HddsConfigKeys.HDDS_SCM_WAIT_TIME_AFTER_SAFE_MODE_EXIT,
+        HddsConfigKeys.HDDS_SCM_WAIT_TIME_AFTER_SAFE_MODE_EXIT_DEFAULT,
+        TimeUnit.MILLISECONDS);
+    this.legacyReplicationManager = new LegacyReplicationManager(
+        conf, containerManager, containerPlacement, eventPublisher,
+        scmContext, nodeManager, scmhaManager, clock, moveTable);
+
+    // register ReplicationManager to SCMServiceManager.
+    serviceManager.register(this);
+
+    // start ReplicationManager.
+    start();
+  }
+
+  /**
+   * Starts Replication Monitor thread.
+   */
+  @Override
+  public synchronized void start() {
+    if (!isRunning()) {
+      LOG.info("Starting Replication Monitor Thread.");
+      running = true;
+      metrics = ReplicationManagerMetrics.create(this);
+      legacyReplicationManager.setMetrics(metrics);
+      replicationMonitor = new Thread(this::run);
+      replicationMonitor.setName("ReplicationMonitor");
+      replicationMonitor.setDaemon(true);
+      replicationMonitor.start();
+    } else {
+      LOG.info("Replication Monitor Thread is already running.");
+    }
+  }
+
+  /**
+   * Returns true if the Replication Monitor Thread is running.
+   *
+   * @return true if running, false otherwise
+   */
+  public boolean isRunning() {
+    if (!running) {
+      synchronized (this) {
+        return replicationMonitor != null
+            && replicationMonitor.isAlive();
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Stops Replication Monitor thread.
+   */
+  public synchronized void stop() {
+    if (running) {
+      LOG.info("Stopping Replication Monitor Thread.");
+      running = false;
+      legacyReplicationManager.clearInflightActions();
+      metrics.unRegister();
+      notifyAll();
```

Comment:
This `notifyAll()` should probably be `replicationMonitor.interrupt()`. I think it works fine like this, because of the synchronized, but I recently came across the `stop()` method in another service not stopping the service quickly. as the notify was issued when the thread was running and not waiting.

If we call interrupt, then it will interrupt the thread when it is sleeping, or it will set the interrupt flag and it will be interrupted when it stops running and tries to `wait()`.

Reply:
this is a good point , i have updated this patch , please take a look!